### PR TITLE
Add implementation of `now`, `sysdate`, `localtime` and similar functions

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Signed-off-by: Yury Fridlyand <yuryf@bitquilltech.com>

## Description
New functions added:
* now()
* current_timestamp()
* localtimestamp()
* localtime()
* sysdate()
* curtime()
* current_time()
* curdate()
* current_date()

New signatures:
* `now()` : `datetime`
* `current_timestamp()` : `datetime`
* `current_timestamp` : `datetime`
* `localtimestamp()` : `datetime`
* `localtimestamp` : `datetime`
* `localtime()` : `datetime`
* `localtime` : `datetime`
* `sysdate()` : `datetime`
* `sysdate(int)` : `datetime`
* `curtime()` : `time`
* `current_time()` : `time`
* `current_time` : `time`
* `curdate()` : `date`
* `current_date()` : `date`
* `current_date` : `date`

<details>
<summary>Not added signatures</summary>

* `now(int)` : `datetime`
* `current_timestamp(int)` : `datetime`
* `localtimestamp(int)` : `datetime`
* `localtime(int)` : `datetime`
* `curtime(int)` : `time`
* `current_time(int)` : `time`
</details>

## Tests

<details>
<summary><b>SQL</b></summary>

```sql
select now(), current_timestamp(), localtimestamp(), localtime(), sysdate(), sysdate(3), curdate(), curtime(), current_time(), current_date(), CURRENT_DATE, CURRENT_TIME, CURRENT_TIMESTAMP, LOCALTIME, LOCALTIMESTAMP;
```

```
fetched rows / total rows = 1/1
-[ RECORD 1 ]-------------------------
now()                   | 2022-07-20 20:15:25.262409
current_timestamp()     | 2022-07-20 20:15:25.262409
localtimestamp()        | 2022-07-20 20:15:25.262409
localtime()             | 2022-07-20 20:15:25.262409
sysdate()               | 2022-07-20 20:15:26.468210
sysdate(3)              | 2022-07-20 20:15:26.468000
curdate()               | 2022-07-20
curtime()               | 20:15:26.469603
current_time()          | 20:15:26.469746
current_date()          | 2022-07-20
CURRENT_DATE            | 2022-07-20
CURRENT_TIME            | 20:15:26.469869
CURRENT_TIMESTAMP       | 2022-07-20 20:15:25.262409
LOCALTIME               | 2022-07-20 20:15:25.262409
LOCALTIMESTAMP          | 2022-07-20 20:15:25.262409
```
</details>

<details>
<summary><b>PPL</b></summary>

```sql
source=calcs | top 1 zzz | eval `now()` = now(), `current_timestamp()` = current_timestamp(), `localtimestamp()` = localtimestamp(), `localtime()` = localtime(), `sysdate()` = sysdate(), `sysdate(3)` = sysdate(3), `curdate()` = curdate(), `curtime()` = curtime(), `current_time()` = current_time(), `current_date()` = current_date(), `CURRENT_DATE` = CURRENT_DATE, `CURRENT_TIME` = CURRENT_TIME, `CURRENT_TIMESTAMP` = CURRENT_TIMESTAMP, `LOCALTIME` = LOCALTIME, `LOCALTIMESTAMP` = LOCALTIMESTAMP;
```

```
fetched rows / total rows = 1/1
-[ RECORD 1 ]-------------------------
zzz                     | f
now()                   | 2022-07-20 20:17:22.451338
current_timestamp()     | 2022-07-20 20:17:22.451338
localtimestamp()        | 2022-07-20 20:17:22.451338
localtime()             | 2022-07-20 20:17:22.451338
sysdate()               | 2022-07-20 20:17:22.760505
sysdate(3)              | 2022-07-20 20:17:22.760000
curdate()               | 2022-07-20
curtime()               | 20:17:22.760584
current_time()          | 20:17:22.760626
current_date()          | 2022-07-20
CURRENT_DATE            | 2022-07-20
CURRENT_TIME            | 20:17:22.760693
CURRENT_TIMESTAMP       | 2022-07-20 20:17:22.451338
LOCALTIME               | 2022-07-20 20:17:22.451338
LOCALTIMESTAMP          | 2022-07-20 20:17:22.451338
```
</details>

## Implementation details
**`NOW` and `SYSDATE` difference**
According to `MySQL` standard
> [NOW()](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_now) returns a constant time that indicates the time at which the statement began to execute.

> [SYSDATE()](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_sysdate) returns the time at which it executes.

I had to save and then to pick time when plugin receives a query. `QueryContext` was chosen to store this unless a better candidate introduced.

**Parser update**
1. `datetimeConstantLiteral` was added to `SQL` and `PPL` parsers. It lists pseudo-constants which actually invoke corresponding functions without `()`. So `CURRENT_DATE` is shortcut for `CURRENT_DATE()` and so on.
2. `AstExpressionBuilder` were updated to call a function when `datetimeConstantLiteral` met.
3. `AstExpressionBuilder` in `PPL` is updated a bit by adding `visitFunction` method and removing some duplicated code. Copied from `AstExpressionBuilder` in `SQL`.
4. `ANTLR` rules were copied from `SQL` to `PPL`.
5. `ConstantFunction` entity added to represent functions which values should be cached in scope of the current query. Cache is stored in `AnalysisContext`. This includes changes in:
  * `SQL` parser
  * `PPL` parser
  * `ExpressionAnalyzer`
6. All functions are cached (are `ConstantFunction`) except:
  * `sysdate`
  * `curdate`
  * `current_date`
7. Since `now()` and `now(int)` are different functions, they can't use common cache. To avoid these functions being executed more than once, `func(int)` signatures of them were removed.

Please see PR https://github.com/Bit-Quill/opensearch-project-sql/pull/113 which adds caching mechanism.

**Doctests**
Doctests for new functions always fail, because test validate exact string match for result. I added docs, but disabled doctests for them.
 
## Issues Resolved
- #722 
- #93
- #46
- #260

## See also
* https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html
* https://github.com/opensearch-project/sql/pull/171
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by MitchellGale-BitQuill — DISMISSED** [review]

**Review by forestmvey — DISMISSED** [review]

**Review by dai-chen — DISMISSED** [review]
Thanks for the changes!

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 1]
```diff
@@ -84,6 +90,89 @@ public void register(BuiltinFunctionRepository repository) {
...
+    repository.register(current_time());
+    repository.register(curdate());
+    repository.register(current_date());
+  }
+
+  /**
+   * NOW() returns a constant time that indicates the time at which the statement began to execute.
+   */
+  private LocalDateTime now(@Nullable Integer fsp) {
+    return formatLocalDateTime(QueryContext::getProcessingStartedTime, fsp);
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:109]: I think queryContext is not proper way to store startedTime. My concern is contextContext is init for each rest request, but in future, if we plan to support scheduled query execution, which required execute query at background. 
My suggestions is rewrite now as literal value in AstBuilder. which means
1. Now is only executed one time.
2. SysDate is executed multiple times.

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:109]: Please, see rework in https://github.com/Bit-Quill/opensearch-project-sql/pull/113

Comment (penghuo) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:109]: looks good.

**Comment by MaxKsyunz** [comment]
@Yury-Fridlyand please update the target branch from `main` to `2.x`

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java]: Cool! All these changes are here, please re-review.

cc: @penghuo @dai-chen

File: sql/src/main/antlr/OpenSearchSQLParser.g4 [hunk 2]
```diff
@@ -414,7 +431,7 @@ legacyRelevanceFunctionName
     ;
 
 functionArgs
-    : functionArg (COMMA functionArg)*
+    : (functionArg (COMMA functionArg)*)?
```
Comment (dai-chen) [inline sql/src/main/antlr/OpenSearchSQLParser.g4:434]: Is this change necessary? Because I see all others use `functionArgs?` already.

File: core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -169,6 +170,19 @@ public Expression visitRelevanceFieldList(RelevanceFieldList node, AnalysisConte
...
+  @Override
+  public Expression visitConstantFunction(ConstantFunction node, AnalysisContext context) {
+    var valueName = node.getFuncName();
+    if (context.getConstantFunctionValues().containsKey(valueName)) {
+      return context.getConstantFunctionValues().get(valueName);
+    }
+
+    var value = visitFunction(node, context);
+    value = DSL.literal(value.valueOf(null));
+    context.getConstantFunctionValues().put(valueName, value);
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java:182]: Just to confirm, the result of first constant function call will be "cached" and reused by later same function call. So `now()` and `curdate()` may be different, right?

**Review by MitchellGale-BitQuill — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java (hunk 3 above)
Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/analysis/ExpressionAnalyzer.java:182]: That is correct.  `sysdate` returns the date when the `sysdate` _operation_ is executed. `now` returns the date at the start of execution of the _query_.

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 4]
```diff
@@ -85,6 +89,84 @@ public void register(BuiltinFunctionRepository repository) {
     repository.register(to_days());
     repository.register(week());
     repository.register(year());
+
+    repository.register(now());
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:93]: nit: should have inserted these alphabetically into the above list

**Review by acarbonetto — DISMISSED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 4 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:93]: I'll update in next PR

File: sql/src/main/antlr/OpenSearchSQLParser.g4 (hunk 2 above)
Comment (Yury-Fridlyand) [reply sql/src/main/antlr/OpenSearchSQLParser.g4:434]: The goal is to have the same behavior in SQL and PPL. This fix makes `FunctionArgsContext` always not null. In case if there is no args, it contains an empty list.
See changes in `AstExpressionBuilder`s: https://github.com/Bit-Quill/opensearch-project-sql/blob/integ-datetime-now/ppl/src/main/java/org/opensearch/sql/ppl/parser/AstExpressionBuilder.java#L272

Comment (dai-chen) [reply sql/src/main/antlr/OpenSearchSQLParser.g4:434]: Okay. My question was with this change, is all question mark still required, one example: https://github.com/Bit-Quill/opensearch-project-sql/blob/integ-datetime-now/sql/src/main/antlr/OpenSearchSQLParser.g4#L303

Comment (Yury-Fridlyand) [reply sql/src/main/antlr/OpenSearchSQLParser.g4:434]: Thank you, I removed them from SQL, PPL already don't have them. See fix in bd79364b.

**Review by dai-chen — APPROVED** [review]
LGTM. Thanks for the changes!

**Review by acarbonetto — APPROVED** [review]

**Review by MaxKsyunz — APPROVED** [review]
